subsys/bluetooth/host/iso.c

```diff
@@ -705,9 +705,32 @@ void bt_iso_recv(struct bt_conn *iso, struct net_buf *buf, uint8_t flags)
 #endif /* CONFIG_BT_ISO_UNICAST) || defined(CONFIG_BT_ISO_SYNC_RECEIVER */
 
 #if defined(CONFIG_BT_ISO_UNICAST) || defined(CONFIG_BT_ISO_BROADCASTER)
+static uint16_t iso_chan_max_data_len(const struct bt_iso_chan *chan,
+				      uint32_t ts)
+{
+	uint16_t max_data_len;
+
+	if (chan->qos->tx == NULL) {
+		return 0;
+	}
+
+	max_data_len = chan->qos->tx->sdu;
+
+	if (ts == BT_ISO_TIMESTAMP_NONE) {
+		max_data_len = MIN(bt_dev.le.iso_mtu - BT_HCI_ISO_DATA_HDR_SIZE,
+				   max_data_len);
+	} else {
+		max_data_len = MIN(bt_dev.le.iso_mtu - BT_HCI_ISO_TS_DATA_HDR_SIZE,
+				   max_data_len);
```

Comment:
I always mix up MTU, PDU and SDU sizes, hence correct me if I'm wrong. But I think MTU maps to PDU, and you can have more PDU's to form a SDU. Hence I don't think this check is needed - as max_data_len refers to the SDU and not the PDU. 
What we need to check is that the SDU do not exceed the ISO_buffer_size * ISO_buffer_count from the read_buffer_size_V2 HCI command. To my knowledge there is no way for a controller to inform the host about it not being able to handle SDU's above certain size should it be smaller than the buffers it have.
If I got this detail wrong, please ping me and I'll approve.

Reply:
There is actually another requirement.
IMO the controller ISO buffers(bt_hci_rp_le_read_buffer_size_v2) also needs to be able to hold BN SDU's - hence if BN is 3 and SDU is 300 and the controller iso_max_len is 256 you would need the controller to have 6 ISO TX buffer entries.
`BN * ceil(SDU/iso_max_len)` needs to be <= `iso_max_num`

Reply:
> I always mix up MTU, PDU and SDU sizes, hence correct me if I'm wrong. But I think MTU maps to PDU, and you can have more PDU's to form a SDU. Hence I don't think this check is needed - as max_data_len refers to the SDU and not the PDU.

Using the term MTU here is a bit odd, as it does correspond to anything in the spec. However it is currently used to cover the `ISO_SDU_Length` of the HCI ISO Data Packet: 
![image](https://user-images.githubusercontent.com/2520413/184604937-4cf8e2a3-3b67-4873-af51-54507fe6f7cf.png)


i.e. the maximum size of the actual data ISO data. That is why, when allocating the buffer, one should use `BT_ISO_SDU_BUF_SIZE` to also handle the ISO Header (Connection_Handle and ISO_Data_Load_Length)  and the ISO SDU header (Time_Stamp, Packet_Sequence_Number, ISO_SDU_Length and the Packet_Status_Flag). 

So we have SDU_Size (BT_ISO_SDU_BUF_SIZE) = MTU size (CONFIG_BT_ISO_TX_MTU) + 12

The PDU size is the maximum size the controller supports sending over air. This value is limited to maximum 251 (whereas SDU size is limited to 4095). 

The `bt_dev.le.iso_mtu` (perhaps misleading name) is then the **buffer** size read from the controller. This may be different from the PDU size, and is used to determine how data is (and if) fragmented when sending ISO data over HCI to the controller. The check your commented on here should actually be removed, and instead be replaced by a check to see if SDU > bt_dev.le.iso_mtu * bt_dev.le.iso_pkts.limit, to determine if is possible to send the data.

My understanding is that we should not worry about BN; this is a controller thing that is unrelated to how the host sends data, or how the buffers are used.

Reply:
It is correct BN is a controller thing, but if BN=3 for your transport and there is not enough buffers to hold data for a full ISO event inside the controller I fear it will not be able to transmit the data (as the host cannot deliver the data in time).  At least some controllers wants to be able to schedule a full ISO event up front - and will not accept new data arriving within an ISO event.

Reply:
> At least some controllers wants to be able to schedule a full ISO event up front - and will not accept new data arriving within an ISO event.

That's a good point. I would have expected that a controller would not attempt to send any data until the entire SDU has been received from the host (i.e. either a `BT_ISO_SINGLE` or a `BT_ISO_END` has been received), but that may not be the case (haven't checked/read on this).